### ISSUE TITLE
allOf is not properly merged

### ISSUE THREAD

**Issue by shadowhand** [cebe/php-openapi#31]
Given a schema that uses `allOf` to combine two different `object` schemas, the properties of the two schemas are not combined:

```json
{
  "components": {
    "schemas": {
      "identifier": {
        "type": "object",
        "properties": {
           "id": {"type": "string"}
        }
      },
      "person": {
        "allOf": [
          {"$ref": "#/components/schemas/identifier"},
          {
            "type": "object",
            "properties": {
              "name": {
                "type": "string"
              }
            }
          }
        ]
      }
    }
  }
}
```

Not 100% sure that schema validates, but it should be a good start.

**Comment by cebe** [comment]
Hi, thanks for the report, could you elaborate on how you read the schema and what you expect in the output. I added a test for your Schema (b9eddbd) and as far as I see everything is loaded correctly.

**Comment by shadowhand** [comment]
I believe the resolved reference should include both the `id` and `name` properties; something like:

```
assert($resolvedRef->type === 'object');
assert($resolvedRef->properties->id->type === 'string');
assert($resolvedRef->properties->name->type === 'string');
```

When I tried to use https://github.com/lezhnev74/openapi-psr7-validator it gave me errors with any schema that used `allOf`, stating that it couldn't find the (combined) properties.

**Comment by cebe** [comment]
The purpose of `php-openapi` is to read and write OpenAPI description files as they are, so you get the raw allOf element with its sub-schemas only. For schema validation you need to use other tools (at least now this is not implemented). The validation of the request schema against the API description has to be made by `lezhnev74/openapi-psr7-validator` which should validate the request based on allOf semantics.

**Comment by cebe** [comment]
ping @lezhnev74

**Comment by lezhnev74** [comment]
Hey! Ok, this looks like a problem with resolving references. Let me try to write a failing test first.

**Comment by lezhnev74** [comment]
I've added this test which validates a response body against a schema with a reference:
- [a test](https://github.com/lezhnev74/openapi-psr7-validator/blob/57c05ec2961c6b5e24faa62247ae753906955c80/tests/PSR7/ValidateResponseTest.php#L26)
- [a schema](https://github.com/lezhnev74/openapi-psr7-validator/blob/57c05ec2961c6b5e24faa62247ae753906955c80/tests/stubs/api.yaml#L133) and [a reference](https://github.com/lezhnev74/openapi-psr7-validator/blob/57c05ec2961c6b5e24faa62247ae753906955c80/tests/stubs/schemas.yaml#L19)

The validator builds a schema in the memory and resolves all the references automatically.
You can access resolved schema like this:
```php
$schema = $validator->getSchema();
$allOf = $schema->paths['/ref']->post->responses['200']->content['application/json']->schema->allOf;
var_dump($allOf[0]->properties['age']->type); // "integer"
var_dump($allOf[1]->properties['name']->type); // "string"
```

**Comment by lezhnev74** [comment]
@cebe as I understand, one can resolve references manually like this:
```php
use cebe\openapi\Reader;
use cebe\openapi\ReferenceContext;
use function realpath;


$filePath = "./schema.yaml";
$schema = Reader::readFromYamlFile($filePath);
$schema->resolveReferences(new ReferenceContext($schema, realpath($filePath)));
```

**Comment by cebe** [comment]
`resolveReferences()` is called automatically by default, so the reference should not be a problem:

`$resolveReferences` parameter defaults to true:

https://github.com/cebe/php-openapi/blob/13b91759f9daccd90729e1c246bab791f13e59c8/src/Reader.php#L93-L100

**Comment by cebe** [comment]
Hi, version [1.3.0](https://github.com/cebe/php-openapi/releases/tag/1.3.0) fixes a lot of issues with references, please try if this is solved by the changes.

**Comment by cebe** [comment]
Closing this issue for now, please comment if the issue persists. I'll reopen it then.

**Comment by hotmeteor** [comment]
@cebe We're still seeing this issue. 

`allOf` components are not resolving correctly.

You can see the attached issue here: https://github.com/hotmeteor/spectator/issues/38

**Comment by cebe** [comment]
@hotmeteor I see https://github.com/hotmeteor/spectator/issues/38#issuecomment-824249505 is closed now, is there still something to fix in cebe/php-openapi or was this only a problem in your tool?

**Comment by hotmeteor** [comment]
No, there's still an issue. I created a workaround in my tool, but I'd prefer to see things merging as expected. This is the example schema that we're testing with: https://github.com/hotmeteor/spectator/blob/master/tests/Fixtures/Components.v1.json

**Comment by hotmeteor** [comment]
@cebe I forgot I had reported this issue, so I had created another ticket here: https://github.com/cebe/php-openapi/issues/120

It can be closed, but it does contain some useful information for testing.

**Comment by SOHELAHMED7** [comment]
This issue is fixed in https://github.com/cebe/php-openapi/pull/208

Until merged, this PR can be 'composer required' as documented at https://github.com/php-openapi/yii2-openapi/blob/master/CONTRIBUTING.md#use-pr-of-your-own-fork-of-this-library-in-your-project-to-check-new-changes

Usage:

```php
$openApi = Reader::readFromYamlFile(__DIR__ . '/path-to-spec.yml', \cebe\openapi\spec\OpenApi\OpenApi::class, true, true);
```
